HARK/ConsumptionSaving/ConsIndShockModel_AgentSolve.py

```diff
@@ -0,0 +1,2377 @@
+    ConsumerSolution of this class, all of which applies here.
+
+    Parameters
+    ----------
+    DiscFac : float
+        Intertemporal discount factor for future utility.
+    LivPrb : float
+        Survival probability; likelihood of being alive at the beginning of
+        the next period.
+    CRRA : float
+        Coefficient of relative risk aversion.
+    Rfree : float
+        Risk free interest factor on end-of-period assets.
+    PermGroFac : float
+        Expected permanent income growth factor at the end of this period.
+    """
+    __doc__ += ConsumerSolution.__doc__
+
+    def __init__(self, *args,
+                 CRRA=2., DiscFac=0.96, LivPrb=1.0, Rfree=1.03, PermGroFac=1.0,
+                 **kwds):
+        ConsumerSolution.__init__(self, *args, **kwds)
+
+        self.Pars.DiscFac = DiscFac
+        self.Pars.LivPrb = LivPrb
+        self.Pars.CRRA = CRRA
+        self.Pars.Rfree = Rfree
+        self.Pars.PermGroFac = PermGroFac
+
+        # These have been moved to Bilt to declutter whiteboard:
+        del self.hNrm
+        del self.vPfunc
+        del self.vPPfunc
+        del self.MPCmax
+        del self.MPCmin
+
+        self.Bilt.transitions_possible = self.define_transitions_possible()
+
+    def define_transitions_possible(self):
+        """
+        Construct dictionary containing the possible transition equations.
+
+        Returns
+        -------
+        possible_transitions : dict
+            Names and definitions of the transitions possible within this
+        solution object or as an exit from it to the beginning of the next
+        object.
+
+        """
+        # Working backwards from End of Problem
+
+        # Later we can select among the allowed transitions
+        # First, for model with shocks at Beginning of Problem/Period (BOP):
+        chosen_to_next_BOP = \
+            {'kNrm': 'kNrm = aNrm'}  # k_{t+1} = a_{t}
+
+        choice_to_chosen = \
+            {'aNrm': 'aNrm = mNrm - cNrm'}  # a_{t} = m_{t} - c_{t}
+
+        BOP_to_choice = {
+            'RNrm': 'RNrm = Rfree / (PermGroFac * permShk)',
+            'bNrm': 'bNrm = kNrm * RNrm',
+            'yNrm': 'yNrm = tranShk',
+            'mNrm': 'mNrm = bNrm + yNrm'}
+
+        # Now, for model with shocks at End of Problem/Period (EOP)
+
+        chosen_to_next_choice = \
+            {'kNrm': 'kNrm = aNrm',
+             'RNrm': 'RNrm = Rfree / (PermGroFac * permShk)',
+             'bNrm': 'bNrm = kNrm * RNrm',  # b_{t} = k_{t} RNrm_{t}
+             'yNrm': 'yNrm = tranShk',  # y_{t} = \tranShk_{t}
+             'mNrm': 'mNrm = bNrm + yNrm'}
+
+        # choice_to_chosen need not be redefined b/c same as defined above
+
+        possible_transitions = \
+            {'chosen_to_next_BOP': chosen_to_next_BOP,
+             'chosen_to_next_choice': chosen_to_next_choice,
+             'choice_to_chosen': choice_to_chosen,
+             'BOP_to_choice': BOP_to_choice
+             }
+
+        return possible_transitions
```

Comment:
a lot of structure is implied by the code here.
I wonder if it would be possible to require that the equations get laid out, and then have the structure (such as the timing of the shocks) be inferred from the model specification.
That's more along the lines of what I'm working on in #865